label_maker/utils.py

```diff
@@ -86,6 +87,46 @@ def get_tile_tif(tile, imagery, folder, imagery_offset):
 
     return tile_img
 
+def get_tile_wms(tile, imagery, folder, imagery_offset):
+    """
+    Read a WMS endpoint with query parameters corresponding to a TMS tile
+
+    Converts the tile boundaries to the spatial reference system (SRS) specified
+    by the WMS query parameter.
+    """
+    query_dict = parse_qs(imagery.lower())
+    image_format = query_dict['format'][0].split('/')[1]
+    wms_srs = query_dict['srs'][0]
+
+    bound = bounds(*[int(t) for t in tile.split('-')])
+    p1 = Proj({'init': 'epsg:4326'})
+    p2 = Proj({'init': wms_srs})
+
+    # project tile boundaries from lat/lng to WMS SRS
+    tile_ll_proj = transform(p1, p2, bound.west, bound.south)
+    tile_ur_proj = transform(p1, p2, bound.east, bound.north)
+    bbox = tile_ll_proj + tile_ur_proj
+
+    wms_url = imagery.replace('{bbox}', ','.join([str(b) for b in bbox]))
+    r = requests.get(wms_url)
+    tile_img = op.join(folder, '{}.{}'.format(tile, image_format))
+    with open(tile_img, 'wb') as w:
+        w.write(r.content)
+    return tile_img
+
+
 def is_tif(imagery):
     """Determine if an imagery path has a valid tif extension"""
     return op.splitext(imagery)[1].lower() in ['.tif', '.tiff', '.vrt']
+
+def is_wms(imagery):
+    """Determine if an imagery path is a WMS endpoint"""
+    return '{bbox}' in imagery
+
+def get_image_function(imagery):
+    """Return the correct image downloading function based on the imagery string"""
+    if is_tif(imagery):
+        return get_tile_tif
+    if is_wms(imagery):
+        return get_tile_wms
```

Comment:
Add something like
```
else:
    raise ValueError('Imagery type not understood')
```

Reply:
The current `else` statement is the fallback function (there isn't a good test for non valid imagery string yet)

Reply:
Where is the current `else` statement?

Reply:
We won't really know if the imagery is bad or not until the downloading function fails

Reply:
Okay